### PR TITLE
[Datamodules] Merge feature branch

### LINKED ISSUE #681: Datamodules Refactor
Changes to be made to the datamodules [feature branch](https://github.com/openvinotoolkit/anomalib/tree/feature/datamodules) before merging into main.

- [x] Create and implement new design for modules in `anomalib/data`.
- [x] https://github.com/openvinotoolkit/anomalib/issues/646
- [x] Add synthetic validation set functionality.
- [x] Remove `PreProcessor` class and use raw transforms instead.
- [x] https://github.com/openvinotoolkit/anomalib/issues/680
- [x] Use enum for task type
- [x] https://github.com/openvinotoolkit/anomalib/pull/732
- [x] Allow custom val/test split ratio (see https://github.com/openvinotoolkit/anomalib/issues/731)

### PR DESCRIPTION
**Pull request by djdameln**
# Description

Monster PR to finalize the datamodules refactor. 

Most important changes:

- Switch to new design with base classes for Dataset and DataModule.
- Add base classes for video datasets, and implementations for the UCSDped and Avenue datasets.
- Add Detection task type for data loaders and visualizations.
- Add Synthetic anomaly generation for validation and evaluation purposes.
- Deprecate PreProcessor class and use Albumentations transforms directly.

- Fixes #681 

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which refactors the code base)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the [pre-commit style and check guidelines](https://github.com/openvinotoolkit/anomalib/blob/main/CONTRIBUTING.md) of this project.
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] I have added a summary of my changes to the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) (not for minor changes, docs and tests).

### CODE REVIEW

File: anomalib/config/config.py [hunk 1]
```diff
@@ -109,6 +127,78 @@ def update_multi_gpu_training_config(config: Union[DictConfig, ListConfig]) -> U
...
+            DeprecationWarning(
+                "The 'split_ratio' parameter is deprecated and will be removed in a future release. Please use "
+                "'test_split_ratio' instead."
+            )
+        )
+        config.dataset.test_split_ratio = config.dataset.split_ratio
+
+    if config.dataset.get("test_split_mode") == TestSplitMode.NONE and config.dataset.get("val_split_mode") in [
+        ValSplitMode.SAME_AS_TEST,
+        ValSplitMode.FROM_TEST,
```
Comment (samet-akcay) [inline anomalib/config/config.py:188]:
```python

  if config.dataset.get("test_split_mode") == TestSplitMode.NONE and config.dataset.get("val_split_mode") in (
        ValSplitMode.SAME_AS_TEST,
        ValSplitMode.FROM_TEST,
    ):
```

File: anomalib/data/base/datamodule.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+
+class AnomalibDataModule(LightningDataModule, ABC):
+    """Base Anomalib data module.
+
+    Args:
+        train_batch_size (int): Batch size used by the train dataloader.
+        test_batch_size (int): Batch size used by the val and test dataloaders.
+        num_workers (int): Number of workers used by the train, val and test dataloaders.
+        seed (Optional[int], optional): Seed used during random subset splitting.
```
Comment (samet-akcay) [inline anomalib/data/base/datamodule.py:59]: Docstring is missing the following variables.
```python
        val_split_mode: ValSplitMode,
        val_split_ratio: float,
        test_split_mode: Optional[TestSplitMode] = None,
        test_split_ratio: Optional[float] = None,
```

File: anomalib/data/synthetic.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+from anomalib.data.utils import Augmenter, Split, read_image
+
+logger = logging.getLogger(__name__)
+
+
+ROOT = "./.tmp/synthetic_anomaly"
+
+
+def make_synthetic_dataset(
+    source_samples: DataFrame, im_dir: Path, mask_dir: Path, anomalous_ratio: float = 0.5
```
Comment (samet-akcay) [inline anomalib/data/synthetic.py:34]: Just to follow a pattern, maybe we could have one of the two:

```python
source_samples: DataFrame, image_dir: Path, mask_dir: Path, anomalous_ratio: float = 0.5
```

or

```python
source_samples: DataFrame, img_dir: Path, msk_dir: Path, anomalous_ratio: float = 0.5
```

File: anomalib/data/synthetic.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+        # create directories
+        self.im_dir.mkdir()
+        self.mask_dir.mkdir()
+
+        self._cleanup = True  # flag that determines if temp dir is cleaned up when instance is deleted
+        self.setup()
+
+    @classmethod
+    def from_dataset(cls, dataset):
```
Comment (samet-akcay) [inline anomalib/data/synthetic.py:133]: Would it be possible to annotate this classmethod?

File: anomalib/data/ucsd_ped.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
+"""UCSD Pedestrian dataset."""
+
```
Comment (samet-akcay) [inline anomalib/data/ucsd_ped.py:2]: License is missing

File: anomalib/data/ucsd_ped.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+    get_transforms,
+    hash_check,
+    read_image,
+)
+from anomalib.data.utils.video import ClipsIndexer
+
+logger = logging.getLogger(__name__)
+
+
+def make_ucsd_dataset(path: Path, split: Optional[Union[Split, str]] = None):
```
Comment (samet-akcay) [inline anomalib/data/ucsd_ped.py:33]:
```python
def make_ucsd_dataset(path: Path, split: Optional[Union[Split, str]] = None) -> DataFrame:
```

File: anomalib/pre_processing/pre_process.py [hunk 7]
```diff
@@ -8,16 +8,56 @@
...
 
 import albumentations as A
 from albumentations.pytorch import ToTensorV2
 
-from anomalib.data.utils import get_image_height_and_width
-
 logger = logging.getLogger(__name__)
 
 
+def get_image_height_and_width(image_size: Optional[Union[int, Tuple]] = None) -> Tuple[Optional[int], Optional[int]]:
```
Comment (samet-akcay) [inline anomalib/pre_processing/pre_process.py:20]: Once we remove `PreProcessor`, I think we would remove the `pre_process.py` file as well. Do you think if this is the best place for this function?

File: anomalib/utils/cv/connected_components.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+import cv2
+import numpy as np
+import torch
+from kornia.contrib import connected_components
+from torch import Tensor
+
+
+def connected_components_gpu(binary_input: Tensor, num_iterations: int = 1000) -> Tensor:
```
Comment (samet-akcay) [inline anomalib/utils/cv/connected_components.py:13]: How about naming `binary_input` as `image`? This is how `kornia` names this. Also `connected_components_cpu` function too

File: anomalib/data/utils/augmenter.py [hunk 9]
```diff
@@ -36,9 +36,19 @@ class Augmenter:
     Args:
         anomaly_source_path (Optional[str]): Path to a folder of images that will be used as source of the anomalous
         noise. If not specified, random noise will be used instead.
+        p_anomalous (float): Probability that the anomalous perturbation will be applied to a given image.
```
Comment (samet-akcay) [inline anomalib/data/utils/augmenter.py:39]: not related to this change, but I'm not sure if `Augmenter` would actually reveal what the class is doing. Maybe we could rename this class so everyone could easily understand what it does?

File: anomalib/data/utils/split.py [hunk 10]
```diff
@@ -11,76 +11,128 @@
...
-    num_normal_valid_images = int(num_normal_train_images * split_ratio)
+    NONE = "none"
+    FROM_DIR = "from_dir"
+    SYNTHETIC = "synthetic"
 
-    indices_to_split_from_train_set = random.sample(population=normal_train_image_indices, k=num_normal_valid_images)
-    samples.loc[indices_to_split_from_train_set, "split"] = "test"
 
-    return samples
+class ValSplitMode(str, Enum):
```
Comment (samet-akcay) [inline anomalib/data/utils/split.py:43]: it would be a bit verbose, but how about `ValidationSplitMode`? Test split has the full name, maybe we could follow the same convention?

**Review by samet-akcay — CHANGES_REQUESTED** [review]
@djdameln, thanks a lot for this PR. There is huge effort here! Much appreciated!

I have some comments, most of which are minor regarding some naming and stuff. Thanks again!

File: anomalib/data/base/video.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        assert isinstance(self.indexer, ClipsIndexer)
+        return self.indexer.num_clips()
+
+    @property
+    def samples(self):
+        """Get the samples dataframe."""
+        return super().samples
+
+    @samples.setter
+    def samples(self, samples):
```
[the remark this answers is not in the document]
Comment (djdameln) [reply anomalib/data/base/video.py:48]: If I add the type here I'm getting a false positive from mypy. 
```
anomalib/data/base/video.py:51: error: overloaded function has no attribute "fset"  [attr-defined]
Found 1 error in 1 file (checked 3 source files)
```
Not sure how to proceed

File: anomalib/data/utils/augmenter.py (hunk 9 above)
Comment (djdameln) [reply anomalib/data/utils/augmenter.py:39]: I agree, but I also think this class needs some refactor in general. Would it be alright if we address this in a separate PR to speed things up?

File: anomalib/data/utils/split.py (hunk 10 above)
Comment (djdameln) [reply anomalib/data/utils/split.py:43]: Done. Should I also change the parameter name in the config and the variable names accordingly?

File: anomalib/pre_processing/pre_process.py (hunk 7 above)
Comment (djdameln) [reply anomalib/pre_processing/pre_process.py:20]: This is the legacy implementation of this function, which is needed by the `PreProcessor` class. So this is just for backward compatibility, and we can safely remove it once we remove `PreProcessor`. The new function that's actually being used throughout the code is in `anomalib.data.utils.image.get_image_width_and_height`.

Comment (samet-akcay) [reply anomalib/pre_processing/pre_process.py:20]: oh, I see. Thanks for the clarification

File: anomalib/data/utils/split.py (hunk 10 above)
Comment (djdameln) [reply anomalib/data/utils/split.py]: Hmm, I'm not sure actually. `validation_split_mode` and `validation_split_ratio` are pretty verbose, which might be an annoyance for users who want to change these parameters through CLI. Also I feel that for terms starting with `val_split` it is immediately clear that 'val' stands for 'validation'.

What do you think @ashwinvaidya17 ?

Comment (djdameln) [reply anomalib/data/utils/split.py:43]: Let's go with `val_split_mode` then. I've reverted the changes to the enum, because I feel the enum should have the same name as the variable.

**Review by samet-akcay — APPROVED** [review]
thanks, great effort!
